### PR TITLE
Integrate logic from RushTIScheduler

### PR DESCRIPTION
**Pull request by ichermak**

### CODE REVIEW

File: RushTI.py [hunk 1]
```diff
@@ -83,6 +103,175 @@ def decrypt_password(encrypted_password):
...
+    levels = deduce_levels_of_tasks(**tasks)
+    levels_count = len(levels.items())
+    for task_key in tasks.keys():
+        index = 0
+        while index < levels_count:
+            level_key = index
+            level = levels[level_key]
+            if level_key + 1 < levels_count:
+                next_level = levels[level_key + 1]
+                if len(next_level) < int(maximum_workers):
```
Comment (MariusWirtz) [inline RushTI.py]: @ichermak 
why do we do this check ? Not sure if it makes sense 🤔

**Comment by MariusWirtz** [comment]
Closed in favor of #22 

Hi @ichermak ,

I squashed your changes into one commit and added another commit, with my changes, on top.
Please review my changes here #22 

I think I was able to comprehend everything, except for the `rearrange_tasks_in_levels` method. Didn't figure out why it was necessary.
I also addded some test cases to the project. Currently only the functionality from the `deduce_levels_of_tasks` method is covered.
Apart from these two points my changes are mostly related to style and readability (e.g. replacing dictionaries with classes and introducing an enumeration for the execution type.)

**Comment by ichermak** [comment]
Hi @MariusWirtz ,

RushTiScheduler wrappes tow ideas : 

1. The ability to specify dependencies within the tasks file: `deduce_levels_of_tasks` method guarantees those dependencies during execution. 

1. An optimization of the execution plan:  This is done by reordering the tasks within the levels regarding the maximum workers. That's the role of `rearrange_tasks_in_levels` method. This part is note necessary but an interesting feature in my opinion. I have some examples show the enhancement provided to execution time. I can share If you are interested.

**Comment by MariusWirtz** [comment]
Hi @ichermak ,

OK. Please share the sample. I think that will help me to grasp the idea!
If you find time you could implement the samples in the shape of tests. I did it for the point 1. here:
tasks files:
https://github.com/cubewise-code/rushti/tree/feature/rush-ti-scheduler-refactoring/tests/resources
check expected outcome:
https://github.com/cubewise-code/rushti/blob/feature/rush-ti-scheduler-refactoring/tests/tests.py

Otherwise you can also just drop a file and the expected outcome to this thread.

Once I understand, I would uncomment the function call for point 2. then we can merge in the changes.
Thanks for the contribution!

**Comment by ichermak** [comment]
Hi @MariusWirtz ,

I’m not sure I can do this by implementing a test in tests.py. 
You can compare time execution using the sample [Tasks_type_optimized.txt](https://github.com/cubewise-code/rushti/blob/feature/rush-ti-scheduler-refactoring/Tasks_type_optimized.txt) : 

- Case 1:	With VS Without `rearrange_tasks_in_levels` method and with maximum_workers = 2

- Case 2:	With VS Without `rearrange_tasks_in_levels` method and with maximum_workers = 3

You will get a better result using the `rearrange_tasks_in_levels`.

**Comment by MariusWirtz** [comment]
Hi @ichermak ,

Thank you. I understand the logic now. It makes sense.

I added one minor constraint to the rearrangement logic. 
Now we only push tasks to the next level, if the number of tasks in the current level is greater than (or equal to) the `maximum_workers`. Does that make sense ?
And I renamed the method to `balance_tasks_among_levels`

If you have no more input about #22, I will accept the MR.

**Comment by ichermak** [comment]
Hi @MariusWirtz, 

That's exactly the idea. 
Just notice that the condition to move a task to the next level must be : if the number of tasks in the current level are greater than the `maximum_workers`. If it's equal, it means we are using all the workers wich is optimal.

**Comment by MariusWirtz** [comment]
Hi @ichermak,

I had the same thought!
Howevery if you test it, you will see that this condition:
`if len(level) >= max_workers >= len(next_level):`
delivers better results than this one:
`if len(level) > max_workers >= len(next_level):`
I suppose this is because of the iterative approach we are using.

Thanks again for your contribution. I will Merge the changes tonight and draft a new release.

**Pull request closed without merging**